src/Functions/sleep.h

```diff
@@ -128,7 +135,19 @@ class FunctionSleep : public IFunction
                         "The maximum sleep time is {} microseconds. Requested: {} microseconds per block (of size {})",
                         max_microseconds, microseconds, size);
 
-                sleepForMicroseconds(microseconds);
+                while (microseconds)
+                {
+                    UInt64 sleep_ms = microseconds;
+                    if (query_status)
+                        sleep_ms = std::min(sleep_ms, /* 1 second */ static_cast<size_t>(1000000));
+
+                    sleepForMicroseconds(sleep_ms);
+                    microseconds -= sleep_ms;
+
+                    if (query_status && query_status->isKilled())
```

Comment:
WDYT about calling `checkTimeLimit()` instead? That way it works both for KILL and `max_execution_time` and it removes the need to throw the exception here (checkTimeLimit does it itself).